pkg/plugin/processor/standalone/wasm_processor.go

```diff
@@ -0,0 +1,258 @@
+	"github.com/tetratelabs/wazero"
+	"github.com/tetratelabs/wazero/api"
+	"github.com/tetratelabs/wazero/imports/wasi_snapshot_preview1"
+)
+
+type wasmProcessor struct {
+	logger zerolog.Logger
+
+	runtime       wazero.Runtime
+	runtimeCancel context.CancelFunc
+
+	// commandCh is used to send commands to
+	// the actual processor (the WASM module)
+	// and through exported host functions.
+	commandCh chan sdk.Command
+	// replyErr is used to communicate replies between
+	// the actual processor (the WASM module) and wasmProcessor
+	replyCh chan sdk.CommandResponse
+	// replyErr is used to communicate errors between
+	// the actual processor (the WASM module) and wasmProcessor
+	replyErr chan error
+	// runModStopped is used to know when the run module stopped
+	runModStopped chan struct{}
+}
+
+func NewWASMProcessor(ctx context.Context, logger zerolog.Logger, wasmPath string) (sdk.Processor, error) {
+	runtimeCtx, runtimeCancel := context.WithCancel(ctx)
+	r := wazero.NewRuntimeWithConfig(
+		runtimeCtx,
+		wazero.NewRuntimeConfig().WithCloseOnContextDone(true),
+	)
+	wasi_snapshot_preview1.MustInstantiate(ctx, r)
+
+	p := &wasmProcessor{
+		logger:        logger,
+		runtime:       r,
+		runtimeCancel: runtimeCancel,
+		commandCh:     make(chan sdk.Command),
+		replyCh:       make(chan sdk.CommandResponse),
+		replyErr:      make(chan error),
+		runModStopped: make(chan struct{}),
+	}
+
+	err := p.exportFunctions(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed exporting processor functions: %w", err)
+	}
+
+	err = p.run(ctx, wasmPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed running WASM module: %w", err)
+	}
+
+	return p, nil
+}
+
+func (p *wasmProcessor) exportFunctions(ctx context.Context) error {
+	// Build a host module, called `env`, which will expose
+	// functions which WASM processor module can use,
+	envBuilder := p.runtime.NewHostModuleBuilder("env")
+	envBuilder.
+		NewFunctionBuilder().
+		WithFunc(p.nextCommand).
+		Export("nextCommand")
+	envBuilder.
+		NewFunctionBuilder().
+		WithFunc(p.reply).
+		Export("reply")
+
+	_, err := envBuilder.Instantiate(ctx)
+	return err
+}
+
+func (p *wasmProcessor) nextCommand(ctx context.Context, m api.Module, ptr, allocSize uint32) uint32 {
+	p.logger.Trace().Msg("getting next command")
+
+	cmd, ok := <-p.commandCh
+	if !ok {
+		p.logger.Info().Msg("command channel closed")
+		return sdk.ErrCodeNoMoreCommands
+	}
+
+	bytes, err := serde.MarshalCommand(cmd)
+	if err != nil {
+		p.logger.Err(err).Msg("failed marshaling command")
+		p.replyErr <- fmt.Errorf("failed marshaling command: %w", err)
+
+		return sdk.ErrCodeFailedGettingCommand
+	}
+
+	return p.write(ctx, m, ptr, allocSize, bytes)
+}
+
+func (p *wasmProcessor) write(_ context.Context, mod api.Module, ptr uint32, sizeAllocated uint32, bytes []byte) uint32 {
+	p.logger.Trace().
+		Int("total_bytes", len(bytes)).
+		Uint32("allocated_size", sizeAllocated).
+		Msgf("writing command to module memory")
```

Comment:
Nice, I like the trace. I'd maybe add the module name as an attribute as well.

Reply:
For the record, all relevant metadata will be in the logger already (processor id, plugin name).